File: app/src/main/java/co/wm21/https/view/adapters/PopularProductAdapter.java
package co.wm21.https.view.adapters;

import static java.lang.Math.round;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.graphics.Paint;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import co.wm21.https.FHelper.ConstantValues;
import co.wm21.https.FHelper.networks.Models.AddToCartModel;
import co.wm21.https.FHelper.networks.Models.ProductModel;
import co.wm21.https.R;
import co.wm21.https.view.activities.MainActivity;
import co.wm21.https.utils.dialog.LoadingDialog;
import co.wm21.https.view.fragments.CartFragment;
import co.wm21.https.helpers.SessionHandler;
import co.wm21.https.presenter.interfaces.OnAddToCartView;
import co.wm21.https.presenter.AddToCartPresenter;

public class PopularProductAdapter extends RecyclerView.Adapter<PopularProductAdapter.ProductViewHolder>implements OnAddToCartView {
    private ArrayList<ProductModel> productList;
    public ItemClickListener listener;
    private Context context;
    SessionHandler sessionHandler;
    private AddToCartPresenter addToCartPresenter;
    View clickView;
    //private ProgressDialog progressDialog;
    LoadingDialog loadingDialog;
    ProgressDialog progressBar;
    private  String selectedColor = "", selectedSize = "";

    public PopularProductAdapter(ArrayList<ProductModel> productList, Context context) {
        this.productList = productList;
        this.context = context;
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ProductViewHolder(LayoutInflater.from(context).inflate(R.layout.layout_item_product, parent, false));
    }

    @SuppressLint({"SetTextI18n", "DefaultLocale"})
    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        ProductModel product = productList.get(position);
        sessionHandler = new SessionHandler(context);
        holder.disLayout.setVisibility(View.GONE);
        if (sessionHandler.isLoggedIn()) {
            holder.rpLayout.setVisibility(View.VISIBLE);
            holder.eshopTV.setVisibility(View.GONE);
            holder.productRPTV.setText(String.format("%s", product.getPoint()));
        } else {
            holder.rpLayout.setVisibility(View.GONE);
            holder.eshopTV.setVisibility(View.VISIBLE);
        }
        double discount=Double.parseDouble(product.getDiscount());
        double sPrice=Double.parseDouble(product.getSprice());

        if(round(discount/sPrice*100)>.49){
            holder.disLayout.setVisibility(View.VISIBLE);
            holder.discountText.setText(round( discount / sPrice * 100) + "% OFF");
        }

        holder.shopName.setText("(" + product.getUploadBy() + ")");
        holder.productName.setText(product.getName());
        holder.previousPrice.setText(String.format("৳ %s", product.getPrice()));
        holder.productPrice.setText(String.format("৳ %s", product.getSprice()));
        // holder.productImage.setImageDrawable(product.getImage());
        Picasso.get().load(ConstantValues.URL+"image/product/small/"+product.getImg()).into(holder.productImage);

        holder.btnAddToCart.setOnClickListener(view -> {
            loadingDialog=new LoadingDialog((Activity) context);
            addToCartPresenter=new AddToCartPresenter(this);
            @SuppressLint("HardwareIds") String val = Settings.Secure.getString(((FragmentActivity) context).getContentResolver(), Settings.Secure.ANDROID_ID);
            String uId = val.replaceAll("[\\D]", "");
            int qty = 1;
            clickView=view;
            addToCartPresenter.AddToCartDataLoad(String.valueOf(product.getSerial()), uId,selectedColor,selectedSize,qty);


        });
    }

    @Override
    public int getItemCount() {
        return productList.size();
    }
    public void setOnClickListener(ItemClickListener listener) {
        this.listener = listener;
    }

    public PopularProductAdapter addOnClickListener(ItemClickListener listener) {
        this.listener = listener;
        return this;
    }

    public class ProductViewHolder extends RecyclerView.ViewHolder {
        ImageView productImage;
        TextView productName, productPrice, previousPrice, btnAddToCart, productRPTV, shopName,eshopTV;
        LinearLayout productRPLayout,rpLayout;
        TextView discountText;
        RelativeLayout disLayout;
        public ProductViewHolder(@NonNull View itemView) {
            super(itemView);
            disLayout = itemView.findViewById(R.id.disLayout);
            discountText = itemView.findViewById(R.id.disText);
            productImage = itemView.findViewById(R.id.productImage);
            eshopTV = itemView.findViewById(R.id.eshopTV);
            productName = itemView.findViewById(R.id.productName);
            productPrice = itemView.findViewById(R.id.productPrice);
            previousPrice = itemView.findViewById(R.id.previousPrice);
            btnAddToCart = itemView.findViewById(R.id.addToCart);
            productRPTV = itemView.findViewById(R.id.productRP);
            productRPLayout = itemView.findViewById(R.id.productRPLayout);
            rpLayout = itemView.findViewById(R.id.rpLayout);
            shopName = itemView.findViewById(R.id.shopNameTV);
            previousPrice.setPaintFlags(previousPrice.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
            // to remove "previousPrice.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG)"
            if (listener != null)
                itemView.setOnClickListener(v -> listener.OnClick(v, getAdapterPosition()));
        }
    }




    @Override
    public void onAddToCartDataLoad(AddToCartModel addToCartModel) {

        String status = addToCartModel.getData();
        String error = String.valueOf(addToCartModel.getError());
        if (error.equals("0") || error.equals("2")) {

            AlertDialog.Builder builder = new AlertDialog.Builder(context);
            ViewGroup viewGroup = clickView.findViewById(android.R.id.content);
            View dialogView = ((FragmentActivity) context).getLayoutInflater().inflate(R.layout.layout_add_to_cart_item, viewGroup, false);
            RelativeLayout ok = dialogView.findViewById(R.id.okBtnId);
            RelativeLayout goToCart = dialogView.findViewById(R.id.goBtnId);
            TextView msg = dialogView.findViewById(R.id.textMsg);
            msg.setText(addToCartModel.getData());
            builder.setView(dialogView);
            AlertDialog alertDialog = builder.create();
            alertDialog.show();
            ok.setOnClickListener(view1 -> {
                alertDialog.dismiss();
            });
            goToCart.setOnClickListener(view1 -> {
                alertDialog.dismiss();

                ((MainActivity) context).switchFragment(new CartFragment(), "CartFragment");
            });

            ((MainActivity) context).getCartItems();

        }
        Snackbar.make(((MainActivity)context).findViewById(R.id.fragmentContainer), status, Snackbar.LENGTH_SHORT).show();


    }

    @Override
    public void onAddToCartStartLoading() {
        loadingDialog.startLoadingAlertDialog();
    }

    @Override
    public void onAddToCartStopLoading() {
       loadingDialog.dismissDialog();
    }

    @Override
    public void onAddToCartShowMessage(String errmsg) {
        Toast.makeText(context, errmsg, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/co/wm21/https/view/adapters/PopularProductAdapter.kt
package co.wm21.https.view.adapters

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.graphics.Paint
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import co.wm21.https.FHelper.ConstantValues
import co.wm21.https.FHelper.networks.Models.AddToCartModel
import co.wm21.https.FHelper.networks.Models.ProductModel
import co.wm21.https.R
import co.wm21.https.helpers.SessionHandler
import co.wm21.https.presenter.AddToCartPresenter
import co.wm21.https.presenter.interfaces.OnAddToCartView
import co.wm21.https.utils.dialog.LoadingDialog
import co.wm21.https.view.activities.MainActivity
import co.wm21.https.view.fragments.CartFragment
import coil.decode.SvgDecoder
import coil.load
import com.google.android.material.snackbar.Snackbar

class PopularProductAdapter(
    private val productList: ArrayList<ProductModel>,
    private val context: Context
) : RecyclerView.Adapter<PopularProductAdapter.ProductViewHolder>(), OnAddToCartView {
    var listener: ItemClickListener? = null
    var sessionHandler: SessionHandler? = null
    private var addToCartPresenter: AddToCartPresenter? = null
    var clickView: View? = null

    //private ProgressDialog progressDialog;
    var loadingDialog: LoadingDialog? = null
    var progressBar: ProgressDialog? = null
    private val selectedColor = ""
    private val selectedSize = ""

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        return ProductViewHolder(
            LayoutInflater.from(
                context
            ).inflate(R.layout.layout_item_product, parent, false)
        )
    }

    @SuppressLint("SetTextI18n", "DefaultLocale")
    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        val product = productList[position]
        sessionHandler = SessionHandler(context)
        holder.disLayout.visibility = View.GONE
        if (sessionHandler!!.isLoggedIn) {
            holder.rpLayout.visibility = View.VISIBLE
            holder.eshopTV.visibility = View.GONE
            holder.productRPTV.text = String.format("%s", product.point)
        } else {
            holder.rpLayout.visibility = View.GONE
            holder.eshopTV.visibility = View.VISIBLE
        }
        val discount = product.discount?.toDouble()?:0.0
        val sPrice = product.sprice?.toDouble()?:0.0

        if (Math.round(discount / sPrice * 100) > .49) {
            holder.disLayout.visibility = View.VISIBLE
            holder.discountText.text = Math.round(discount / sPrice * 100).toString()
        }

        holder.shopName.text = "(" + product.uploadBy + ")"
        holder.productName.text = product.name
        holder.previousPrice.text = String.format("৳ %s", product.sprice)
        holder.productPrice.text = String.format("৳ %s", product.price)

        holder.productImage.load(ConstantValues.imageURL + "image/product/small/" + product.img){
            decoderFactory(SvgDecoder.Factory())
        }

        holder.btnAddToCart.setOnClickListener { view: View? ->
            loadingDialog = LoadingDialog(context as Activity)
            addToCartPresenter = AddToCartPresenter(this)
            @SuppressLint("HardwareIds") val `val` = Settings.Secure.getString(
                (context as FragmentActivity).contentResolver,
                Settings.Secure.ANDROID_ID
            )
            val uId = `val`.replace("[\\D]".toRegex(), "")
            val qty = 1
            clickView = view
            addToCartPresenter!!.AddToCartDataLoad(
                product.serial.toString(),
                uId,
                selectedColor,
                selectedSize,
                qty
            )
        }
    }

    override fun getItemCount(): Int {
        return productList.size
    }

    fun setOnClickListener(listener: ItemClickListener?) {
        this.listener = listener
    }

    fun addOnClickListener(listener: ItemClickListener?): PopularProductAdapter {
        this.listener = listener
        return this
    }

    inner class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var productImage: ImageView = itemView.findViewById(R.id.productImage)
        var productName: TextView = itemView.findViewById(R.id.productName)
        var productPrice: TextView = itemView.findViewById(R.id.productPrice)
        var previousPrice: TextView = itemView.findViewById(R.id.previousPrice)
        var btnAddToCart: TextView = itemView.findViewById(R.id.addToCart)
        var productRPTV: TextView = itemView.findViewById(R.id.productRP)
        var shopName: TextView = itemView.findViewById(R.id.shopNameTV)
        var eshopTV: TextView = itemView.findViewById(R.id.eshopTV)
        var productRPLayout: LinearLayout = itemView.findViewById(R.id.productRPLayout)
        var rpLayout: LinearLayout = itemView.findViewById(R.id.rpLayout)
        var discountText: TextView = itemView.findViewById(R.id.offerPercent)
        var disLayout: LinearLayout = itemView.findViewById(R.id.offerLayout)

        init {
            previousPrice.paintFlags = previousPrice.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            // to remove "previousPrice.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG)"
            if (listener != null) itemView.setOnClickListener { v: View? ->
                listener!!.OnClick(
                    v,
                    adapterPosition
                )
            }
        }
    }


    override fun onAddToCartDataLoad(addToCartModel: AddToCartModel) {
        val status = addToCartModel.data
        val error = addToCartModel.error.toString()
        if (error == "0" || error == "2") {
            val builder = AlertDialog.Builder(context)
            val viewGroup = clickView!!.findViewById<ViewGroup>(android.R.id.content)
            val dialogView = (context as FragmentActivity).layoutInflater.inflate(
                R.layout.layout_add_to_cart_item,
                viewGroup,
                false
            )
            val ok = dialogView.findViewById<RelativeLayout>(R.id.okBtnId)
            val goToCart = dialogView.findViewById<RelativeLayout>(R.id.goBtnId)
            val msg = dialogView.findViewById<TextView>(R.id.textMsg)
            msg.text = addToCartModel.data
            builder.setView(dialogView)
            val alertDialog = builder.create()
            alertDialog.show()
            ok.setOnClickListener { view1: View? ->
                alertDialog.dismiss()
            }
            goToCart.setOnClickListener { view1: View? ->
                alertDialog.dismiss()
                (context as MainActivity).switchFragment(CartFragment(), "CartFragment")
            }

            (context as MainActivity).getCartItems()
        }
        Snackbar.make(
            (context as MainActivity).findViewById(R.id.fragmentContainer),
            status,
            Snackbar.LENGTH_SHORT
        ).show()
    }

    override fun onAddToCartStartLoading() {
        loadingDialog!!.startLoadingAlertDialog()
    }

    override fun onAddToCartStopLoading() {
        loadingDialog!!.dismissDialog()
    }

    override fun onAddToCartShowMessage(errmsg: String) {
        //Toast.makeText(context, errmsg, Toast.LENGTH_SHORT).show()
    }
}
